File: CliqueApp/Helpers/PhoneNumberFormatter.swift
//
//  PhoneNumberFormatter.swift
//  CliqueApp
//
//  Centralized helpers for normalizing and comparing phone numbers plus
//  generating deterministic pseudo emails used for Firebase Auth.
//

import Foundation

struct PhoneNumberFormatter {
    /// Default country code (US) used when users omit a prefix.
    private static let defaultCountryCode = "1"
    
    /// Removes everything except digits.
    static func digitsOnly(from input: String) -> String {
        String(input.filter { $0.isNumber })
    }
    
    /// Canonical representation (used across Firestore)
    /// For US numbers (10-11 digits), trims to 10 digits
    /// For international numbers, keeps full number with country code
    static func canonical(_ input: String) -> String {
        var digits = digitsOnly(from: input)
        
        // Handle US numbers specifically
        if digits.count == 11, digits.hasPrefix(defaultCountryCode) {
            digits = String(digits.dropFirst())
        }
        
        return digits
    }

    /// Returns an E.164 formatted number (e.g. +12175551212) for Firebase SMS APIs.
    /// Handles international numbers by preserving the + prefix if present.
    static func e164(_ input: String) -> String {
        let trimmed = input.trimmingCharacters(in: .whitespacesAndNewlines)
        
        // If already has +, just clean it up
        if trimmed.hasPrefix("+") {
            let normalized = "+" + trimmed.dropFirst().filter { $0.isNumber }
            return normalized
        }
        
        let digits = digitsOnly(from: trimmed)
        if digits.isEmpty {
            return ""
        }
        
        // If it's a 10-digit number, assume US
        if digits.count == 10 {
            return "+\(defaultCountryCode)\(digits)"
        }
        
        // If it's 11 digits starting with 1, it's US/Canada
        if digits.count == 11, digits.hasPrefix(defaultCountryCode) {
            return "+\(digits)"
        }
        
        // For any other number, assume it already includes country code
        return "+\(digits)"
    }
    
    /// Builds E.164 format from separate country code and phone number
    static func e164(countryCode: String, phoneNumber: String) -> String {
        let cleanCountryCode = countryCode.filter { $0.isNumber }
        let cleanNumber = phoneNumber.filter { $0.isNumber }
        
        if cleanNumber.isEmpty {
            return ""
        }
        
        // Remove leading zeros from phone number
        let trimmedNumber = cleanNumber.trimmingCharacters(in: CharacterSet(charactersIn: "0"))
        
        return "+\(cleanCountryCode)\(trimmedNumber)"
    }
    
    /// Generates a deterministic pseudo email for Firebase Auth's email/password provider.
    static func pseudoEmail(for input: String) -> String {
        let canonicalNumber = canonical(input)
        return "\(canonicalNumber)@cliqueapp.phone"
    }
    
    /// Determines if two phone numbers represent the same value.
    static func numbersMatch(_ lhs: String, _ rhs: String) -> Bool {
        let normalizedLhs = canonical(lhs)
        let normalizedRhs = canonical(rhs)
        
        if !normalizedLhs.isEmpty, normalizedLhs == normalizedRhs {
            return true
        }
        
        if normalizedLhs.count != normalizedRhs.count {
            let longer = normalizedLhs.count > normalizedRhs.count ? normalizedLhs : normalizedRhs
            let shorter = normalizedLhs.count > normalizedRhs.count ? normalizedRhs : normalizedLhs
            
            if !shorter.isEmpty, longer.hasSuffix(shorter) {
                return true
            }
        }
        
        return false
    }
}

File: CliqueAndroid/app/src/main/java/com/yallaconnect/app/core/util/PhoneNumberFormatter.kt
package com.yallaconnect.app.core.util

object PhoneNumberFormatter {
    private const val DEFAULT_COUNTRY_CODE = "1"

    fun digitsOnly(input: String): String = input.filter { it.isDigit() }

    fun canonical(input: String): String {
        var digits = digitsOnly(input)
        if (digits.length == 11 && digits.startsWith(DEFAULT_COUNTRY_CODE)) {
            digits = digits.drop(1)
        }
        return digits
    }

    fun e164(raw: String): String {
        val trimmed = raw.trim()
        if (trimmed.startsWith("+")) {
            val clean = trimmed.drop(1).filter { it.isDigit() }
            return if (clean.isBlank()) "" else "+$clean"
        }

        val digits = digitsOnly(trimmed)
        if (digits.isBlank()) return ""
        return when {
            digits.length == 10 -> "+$DEFAULT_COUNTRY_CODE$digits"
            digits.length == 11 && digits.startsWith(DEFAULT_COUNTRY_CODE) -> "+$digits"
            else -> "+$digits"
        }
    }

    fun e164(countryCode: String, phoneNumber: String): String {
        val country = countryCode.filter { it.isDigit() }
        var digits = phoneNumber.filter { it.isDigit() }
        if (digits.startsWith("0")) {
            digits = digits.dropWhile { it == '0' }
        }
        return if (digits.isBlank()) "" else "+$country$digits"
    }

    fun pseudoEmail(input: String): String = "${canonical(input)}@cliqueapp.phone"

    fun numbersMatch(first: String, second: String): Boolean {
        val lhs = canonical(first)
        val rhs = canonical(second)
        if (lhs.isBlank() || rhs.isBlank()) return false
        if (lhs == rhs) return true
        if (lhs.length > rhs.length) return lhs.endsWith(rhs)
        if (rhs.length > lhs.length) return rhs.endsWith(lhs)
        return false
    }
}
